core/shared/src/main/scala/scalaz/zio/ExitResult.scala

```diff
@@ -26,25 +26,34 @@ sealed abstract class ExitResult[+E, +A] extends Product with Serializable { sel
 
   final def failed: Boolean = !succeeded
 
-  final def fold[Z](completed: A => Z, failed: (E, List[Throwable]) => Z, interrupted: List[Throwable] => Z): Z =
+  final def fold[Z](
+    completed: A => Z,
+    failed: (E, List[Throwable]) => Z,
+    interrupted: (List[Throwable], List[Throwable]) => Z,
+    terminated: (Throwable, List[Throwable]) => Z
+  ): Z =
     self match {
-      case Completed(v)  => completed(v)
-      case Failed(e, ts) => failed(e, ts)
-      case Terminated(e) => interrupted(e)
+      case Completed(v)       => completed(v)
+      case Failed(e, ts)      => failed(e, ts)
+      case Interrupted(e, ts) => interrupted(e, ts)
+      case Terminated(t, ts)  => terminated(t, ts)
     }
 
   final def zipWith[E1 >: E, B, C](that: ExitResult[E1, B])(f: (A, B) => C): ExitResult[E1, C] = (self, that) match {
-    case (ExitResult.Completed(a), ExitResult.Completed(b))   => ExitResult.Completed(f(a, b))
-    case (ExitResult.Failed(e, ts), rb)                       => ExitResult.Failed(e, combine(ts, rb))
-    case (ExitResult.Terminated(ts), rb)                      => ExitResult.Terminated(combine(ts, rb))
-    case (ExitResult.Completed(_), ExitResult.Failed(e, ts))  => ExitResult.Failed(e, ts)
-    case (ExitResult.Completed(_), ExitResult.Terminated(ts)) => ExitResult.Terminated(ts)
+    case (ExitResult.Completed(a), ExitResult.Completed(b))       => ExitResult.Completed(f(a, b))
+    case (ExitResult.Failed(e, ts), rb)                           => ExitResult.Failed(e, combine(ts, rb))
+    case (ExitResult.Interrupted(e, ts), rb)                      => ExitResult.Interrupted(e, combine(ts, rb))
+    case (ExitResult.Terminated(t, ts), rb)                       => ExitResult.Terminated(t, combine(ts, rb))
+    case (ExitResult.Completed(_), ExitResult.Failed(e, ts))      => ExitResult.Failed(e, ts)
+    case (ExitResult.Completed(_), ExitResult.Interrupted(e, ts)) => ExitResult.Interrupted(e, ts)
```

Comment:
This is an interesting case. What should be the result of combining `Interrupted` and `Terminated`? Which one "wins"? Left? Right? `Interrupted`? `Terminated`?

The fact any answer is arbitrary (as far as I can see) and that we have to ask this question at all suggests maybe interruption and termination are aspects of a single term.

What if we had something like:

```scala
object ExitResult[+E, +A] {
  ...
  case class Terminated(causes: List[TerminationCause]) extends ExitResult[Nothing, Nothing]
  ...
}
```

Then:

```scala
sealed trait TerminationCause
object TerminationCause {
  final case class Primary(value: Throwable) extends TerminationCause
  final case class Finalizer(value: Throwable) extends TerminationCause
  final case class Interruption(value: Throwable) extends TerminationCause
}
```

Now there's just one obvious way to combine two "terminated" states: `++`.

I think this design could simplify other code.

One problem I see is that there's no type-level guarantee of a primary exception. I'm not sure what implications that has. The other problem is that what does an empty `List` mean? I guess it means interruption without cause, if you rule out the other possibilities (?).

/cc @regiskuckaertz

Reply:
The initial goal from #267 was to distinguish if the fiber we ran was interrupted or terminated. How will we do that in that case? Search for `Interruption` in the list of causes ? What if there are some but they refer to inner fibers and not the one we're interested in?

Reply:
That looks like a great idea. We can encode the list into the ADT to fix both problems?
```scala
case class Primary(value: Throwable) extends TerminationCause
case class Finalizer(value: Throwable) extends TerminationCause
case class Secondary(value: Throwable, primary: TerminationCause) extends TerminationCause
```

Reply:
```
sealed trait TerminationCause { self =>
  def toList: List[Throwable] = self match {
    case Secondary(v, p) => p.toList.map(v :: _).reverse
    case Primary(v) => v :: Nil
    case Finalizer(v) => v :: Nil
  }
}
```

Reply:
> The initial goal from #267 was to distinguish if the fiber we ran was interrupted or terminated.

@ghostdogpr this can still be achieved by replacing `Primary` with `Interruption` and `FatalError` cases, and perhaps carry on `isFatal` and `isInterruption` flags.